crates/codegen/grammar/src/dsl.rs

```diff
@@ -0,0 +1,509 @@
+#[macro_export]
+macro_rules! slang_grammar {
+
+    ( $( $($idents:ident)+ = $value:tt ; )* ) => {
+        $(
+            slang_grammar_top_level_clause! { $($idents)+ $value }
+        )*
+
+        pub trait GrammarConstructor {
+            fn new() -> std::rc::Rc<$crate::Grammar>;
+        }
+
+        impl GrammarConstructor for $crate::Grammar {
+            fn new() -> std::rc::Rc<$crate::Grammar> {
+                let mut grammar = Self {
+                    elements: std::collections::HashMap::new(),
+                    // TODO: capture these from the macro
```

Comment:
are the macro's versions used anywhere else? If not, I suggest removing them from the `language` crate, to have a single source of truth here, until it is finally moved.